lib/Sema/MiscDiagnostics.cpp

```diff
@@ -3449,7 +3449,222 @@ static void diagnoseUnintendedOptionalBehavior(TypeChecker &TC, const Expr *E,
+      SmallVector<Type, 4> destOptionals;
+      auto destValueType =
+        destType->lookThroughAllAnyOptionalTypes(destOptionals);
+
+      if (!destValueType->isAny())
+        return false;
+
+      SmallVector<Type, 4> srcOptionals;
+      srcType->lookThroughAllAnyOptionalTypes(srcOptionals);
+
+      if (srcOptionals.size() > destOptionals.size()) {
+        difference = srcOptionals.size() - destOptionals.size();
+        return true;
+      } else {
+        return false;
+      }
+    }
+
+    /// Returns true iff the collection upcast coercion is an Optional-to-Any
+    /// coercion.
+    bool isOptionalToAnyCoercion(CollectionUpcastConversionExpr::ConversionPair
+                                   conversion) {
+      if (!conversion.OrigValue || !conversion.Conversion)
+        return false;
+
+      auto srcType = conversion.OrigValue->getType();
+      auto destType = conversion.Conversion->getType();
+      return isOptionalToAnyCoercion(srcType, destType);
+    }
+
+    /// Looks through OptionalEvaluationExprs and InjectIntoOptionalExprs to
+    /// find a child ErasureExpr, returning nullptr if no such child is found.
+    /// Any intermediate OptionalEvaluationExprs will be marked as ignored.
+    ErasureExpr *findErasureExprThroughOptionalInjections(Expr *E) {
+      while (true) {
+        if (auto *next = dyn_cast<OptionalEvaluationExpr>(E)) {
+          // We don't want to re-visit any intermediate optional evaluations.
+          IgnoredExprs.insert(next);
+          E = next->getSubExpr();
+        } else if (auto *next = dyn_cast<InjectIntoOptionalExpr>(E)) {
+          E = next->getSubExpr();
+        } else {
+          break;
+        }
+      }
+      return dyn_cast<ErasureExpr>(E);
+    }
+
+    void emitSilenceOptionalAnyWarningWithCoercion(Expr *E, Type destType) {
+      // Treat an IUO destination type as a regular Optional type, as we cannot
+      // suggest a fix-it of e.g 'as Any!'.
+      if (auto baseType = destType->getImplicitlyUnwrappedOptionalObjectType())
+        destType = OptionalType::get(baseType);
+
+      SmallString<16> coercionString;
+      coercionString += " as ";
+      coercionString += destType->getWithoutParens()->getString();
+
+      TC.diagnose(E->getLoc(), diag::silence_optional_to_any,
+                  destType, coercionString.substr(1))
+        .highlight(E->getSourceRange())
+        .fixItInsertAfter(E->getEndLoc(), coercionString);
+    }
+
+    void visitErasureExpr(ErasureExpr *E, OptionalToAnyCoercion coercion) {
+      if (coercion.shouldSuppressDiagnostic())
+        return;
+
+      auto subExpr = E->getSubExpr();
+
+      // Currently we don't produce Optional-as-Any warnings for implicit IUO
+      // to Any coercions (this doesn't take into consideration implicit
+      // coercions such as Any?! to Any?, however; we warn on those).
+      if (subExpr->getType()->getImplicitlyUnwrappedOptionalObjectType())
+        return;
```

Comment:
It would be great if we could remove this, which would enable warnings for implicit IUO to Any coercions, such as shown in the `dontWarnIUOToAnyCoercion` function in the tests (which should really produce warnings). With IUOs being removed from the type system, I believe this should happen at *some* point; but not sure if it's okay to just enable them this PR.

Reply:
I would say feel free to warn about the IUOs as well. It will fall out from my last commit to always generate Optional<T> from '!', but there is no reason not to do so now, and it's really inconsistent that we haven't been emitting the warnings in that case as well.

Reply:
@rudkx Thanks, I'll go ahead and make that change then. It looks like we have 3 tests that make such implicit IUO coercions, two of which look like they can be updated with `as Any` without affecting the test ([1](https://github.com/apple/swift/blob/d23e2bce6274d3b61c9c10a17f97a27da87d7fdf/test/Constraints/casts_objc.swift#L37) & [2](https://github.com/apple/swift/blob/d23e2bce6274d3b61c9c10a17f97a27da87d7fdf/test/Constraints/existential_metatypes.swift#L90)). Though I'm not sure about [the third one](https://github.com/apple/swift/blob/d23e2bce6274d3b61c9c10a17f97a27da87d7fdf/test/Constraints/diag_ambiguities.swift#L52), can it be updated with an explicit `as Any` or `as [Any]`, or would it be best to just add an expected warning?

Reply:
I would add the expected warning on the last one, but yes I think you're probably right that adding the coercions for the others is fine.

Reply:
@rudkx Thanks, I've gone ahead and done that now.